onyxia-api/src/main/java/fr/insee/onyxia/api/dao/universe/CatalogLoader.java

```diff
@@ -66,19 +71,13 @@ private void updateHelmRepository(CatalogWrapper cw) {
                                                     excludedChart ->
                                                             excludedChart.equalsIgnoreCase(
                                                                     entry.getKey())));
+            // For each service, filter the multiple versions if needed then refresh remaining
+            // versions
             repository.getEntries().values().parallelStream()
                     .forEach(
-                            entry -> {
-                                entry.parallelStream()
-                                        .forEach(
-                                                pkg -> {
-                                                    try {
-                                                        refreshPackage(cw, pkg);
-                                                    } catch (CatalogLoaderException
-                                                            | IOException e) {
-                                                        LOGGER.info("Exception occurred", e);
-                                                    }
-                                                });
+                            charts -> {
+                                epurateChartsList(charts, cw);
+                                refreshChartsList(charts, cw);
```

Comment:
I think it would be more readable (and cleaner code) if one did the filtering with the `filter`-method. But I also see this may require some more refactoring. e.g. something like.
```java
repository.getEntries().values().parallelStream()
        .map(charts -> charts.stream().filter(chart -> shouldVersionBeIncluded(chart, cw.getMultipleServicesMode())))
        .forEach(
                chart -> {
                    // must be rewritten to be single chart
                    refreshChartsList(chart, cw);
                });
```

I guess it's the mutation in `epurateChartsList` that I don't like (hence this suggestion).

Reply:
Thing is you can't use filter as you need some context (previous chart for the `skipPatches logic`, number of charts already kept for the `maxNumber` logic). That's why I opted for this mutating curation